File: engine/src/elements/FixedSize.ts
import BoxConstraints from '../BoxConstraints'
import Element from '../Element'
import Size from '../Size'
import Position from '../Position'

interface FixedSizeProps {
	size: Size,
	child: Element<any>
}

class FixedSize extends Element<FixedSizeProps> {
	props: FixedSizeProps

	layout() {
		const { size, child } = this.props
		this.document.layoutElement(child, BoxConstraints.fromSize(size))
		child.size = size
		child.relPosition = new Position({ left: 0, top: 0 })
		return size
	}
	
	paint() {
		this.document.paintElement(this.props.child)
	}
}

export default FixedSize

File: core/src/elements/fixed_size.cpp
#include "fixed_size.hpp"

namespace aardvark::elements {

FixedSize::FixedSize(std::shared_ptr<Element> child, Size size,
                     bool is_repaint_boundary)
    : SingleChildElement(child, is_repaint_boundary), size(size){};

Size FixedSize::layout(BoxConstraints constraints) {
  document->layout_element(child.get(),
                           BoxConstraints::from_size(size, true /* tight */));
  child->size = size;
  child->rel_position = Position{0 /* left */, 0 /* top */};
  return size;
};

void FixedSize::paint(bool is_changed) {
  document->paint_element(child.get());
};

}  // namespace aardvark::elements
